src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/AvailableLocationsService.kt

```diff
@@ -0,0 +1,119 @@
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.client.nomismapping.NomisMappingClient
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.client.prisonapi.PrisonApiClient
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.common.isTimesOverlap
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.config.PrisonRegime
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.Location
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.request.AvailableLocationsRequest
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.response.AvailableLocation
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.response.AvailableLocationsResponse
+import java.time.LocalDate
+import java.time.LocalTime
+
+private const val FIFTEEN_MINUTES = 15L
+
+@Service
+@Transactional(readOnly = true)
+class AvailableLocationsService(
+  private val locationsService: LocationsService,
+  private val bookedLocationsService: BookedLocationsService,
+  private val prisonRegime: PrisonRegime,
+) {
+  companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  /**
+   * This provides a snapshot (point in time) of available rooms at time of calling. When the user attempts to save the
+   * booking the room could have already been taken so may fail.
+   */
+  fun findAvailableLocations(request: AvailableLocationsRequest): AvailableLocationsResponse {
+    val startOfDay = prisonRegime.startOfDay(request.prisonCode!!)
+    val endOfDay = prisonRegime.endOfDay(request.prisonCode)
+    val prisonVideoLinkLocations = getDecoratedLocationsAt(request.prisonCode)
+    val bookedLocations = bookedLocationsService.findBooked(request.prisonCode, request.date!!, prisonVideoLinkLocations)
+    val meetingDuration = request.bookingDuration!!.toLong()
+
+    val availableLocations = buildList {
+      prisonVideoLinkLocations.forEach { location ->
+        // These time adjustments do not allow for PRE and POST meeting times.
+        var meetingStartTime = startOfDay
+        var meetingEndTime = meetingStartTime.plusMinutes(meetingDuration)
+
+        while (meetingStartTime.isBefore(endOfDay)) {
+          if (bookedLocations.isBooked(location, meetingStartTime, meetingEndTime).not() && request.timeSlots!!.any { slot -> slot.isTimeInSlot(meetingStartTime) }) {
+            // TODO need to check against the room decoration/schedules
+            add(
+              AvailableLocation(
+                // TODO is it possible for the description to be null?
+                name = location.description ?: "UNKNOWN",
+                startTime = meetingStartTime,
+                endTime = meetingEndTime,
+                dpsLocationId = location.dpsLocationId,
+                dpsLocationKey = location.key,
+                // TODO populate usage
+                usage = null,
+              ),
+            )
+          }
```

Comment:
This is a great starting point for the structure.. I can already imagine a sub if ....
`
if (roomAllowsByAnyRuleOrSchedule(room, request)) {
    add(AvailableLocations(...))
}
`
And this nicely contains all the complexity of room attributes and schedules into one call.